core/src/main/java/org/opensearch/sql/planner/SerializablePlan.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.sql.planner;
+
+import java.io.Externalizable;
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import org.apache.commons.lang3.NotImplementedException;
+import org.opensearch.sql.executor.pagination.PlanSerializer;
+
+/**
+ * All subtypes of PhysicalPlan which needs to be serialized (in cursor, for pagination feature)
+ * should follow one of the following options.
+ * <ul>
+ *   <li>Both:
+ *     <ul>
+ *       <li>Override both methods from {@link Externalizable}.</li>
+ *       <li>Define a public no-arg constructor.</li>
+ *     </ul>
+ *   </li>
+ *   <li>
+ *     Overwrite {@link #getPlanForSerialization} to return
+ *     another instance of {@link SerializablePlan}.
+ *   </li>
+ * </ul>
+ */
+public interface SerializablePlan extends Externalizable {
+
+  /**
+   * Argument is an instance of {@link PlanSerializer.CursorDeserializationStream}.
+   */
+  @Override
+  default void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
+    throw new NotImplementedException(String.format("`readExternal` is not implemented in %s",
+        getClass().getSimpleName()));
+  }
+
+  /**
+   * Each plan which has as a child plan should do.
+   * <pre>{@code
+   * out.writeObject(input.getPlanForSerialization());
+   * }</pre>
+   */
+  @Override
+  default void writeExternal(ObjectOutput out) throws IOException {
+    throw new NotImplementedException(String.format("`readExternal` is not implemented in %s",
+        getClass().getSimpleName()));
+  }
```

Comment:
I have similar question: if most physical node doesn't support this, does this mean this should be done in external visitor? Or `PlanSerializer` should be the visitor?

Reply:
These functions are coming from `Externalizable` java API. An object to be serialized should implement `Serializable` (with default serialization) or `Externalizable` (with custom serialization).

Reply:
Yes, I had it in `PushDownRequestBuilder`, it was removed in bc8c73fe5.
`ResourceMonitorPlan` implements `SerializablePlan`, but not needed to re-define those methods, because it redefine `getPlanForSerialization` - they aren't called.
I removed it in e745d2db6.